fix: Generic ParamSpec for constructors

jackulau opened the pull request:
## Summary

For #43

This fixes a type checking bug where generic class constructors would lose their type parameters when passed through ParamSpec functions. The type checker now correctly preserves type variables (`@_`) instead of finalizing them to `Unknown`, allowing proper type inference.

## The Issue

When passing a generic class constructor through a function using ParamSpec, the type parameters were incorrectly finalized to `Unknown`. For example:

```rust
def identity[**P, R](x: Callable[P, R]) -> Callable[P, R]:
    return x

class C[T]:
    def __init__(self, x: T) -> None:
        self.x = x

c2 = identity(C)
reveal_type(c2)  # ERROR: (x: Unknown) -> C[Unknown]
```

At runtime, generic class constructors should preserve their type parameters for inference. However, Pyrefly's subset checking in `subset.rs` wasn't instantiating fresh type variables for generic classes, causing them to be finalized to `Unknown` during ParamSpec inference.

## The Fix

I made several changes to handle generic class constructors properly:

1. Modified `CallTarget::Class` in pyrefly/lib/alt/call.rs:16 to use `TargetWithTParams`:
```rust
Class(TargetWithTParams<ClassTarget>),
```

2. Added fresh variable instantiation for generic classes in pyrefly/lib/solver/subset.rs:34:
```rust
let fresh_class = self.type_order.instantiate_fresh_class(&class, self.type_order.solver);
```

3. Added helper methods in pyrefly/lib/solver/type_order.rs:17 for accessing class type parameters and instantiating fresh class instances.

This ensures generic type parameters remain as solver variables (`@_`) that can be unified during inference, rather than being prematurely finalized to `Unknown`.

## Test Plan

Updated the existing test case `test_param_spec_generic_constructor` in pyrefly/lib/test/paramspec.rs:

- Removed the `bug = "Generic class constructors don't work with ParamSpec"` marker
- Changed expected output from `# E: revealed type: (x: Unknown) -> C[Unknown]` to `# E: revealed type: (x: @_) -> C[@_]`
- Added verification line `x: C[int] = c2(1)` to ensure type inference works

Verification:
- Ran test locally:
```bash
cargo test test_param_spec_generic_constructor
```
Result: passes

- Ran all ParamSpec tests to ensure no regressions:
```bash
cargo test paramspec
```
Result: passes

jackulau committed:
Fix Generic ParamSpec with class constructors (#43)

This commit fixes issue #43 where generic class constructors don't preserve
type parameters when used with ParamSpec functions.

Changes:
1. Modified CallTarget::Class to use TargetWithTParams in call.rs,
   allowing generic type parameters to be preserved through the call
   target system
2. Updated Type::Forall handling in call.rs to also set tparams for
   Class targets
3. Improved subset checking in subset.rs to instantiate fresh type
   variables for generic class constructors, allowing the variables
   to be properly unified during ParamSpec inference instead of being
   finalized to Unknown
4. Updated call_graph.rs to handle the new CallTarget::Class structure
5. Added helper methods to type_order.rs for accessing class tparams
   and instantiating fresh class instances

The key fix is in subset.rs where we now call instantiate_fresh_class() to
create fresh solver variables for the generic class's type parameters. These
variables remain as solver variables (represented as @_) that can be unified
during callable inference, instead of being immediately finalized to Unknown.

Test result: Generic class constructors now preserve type parameters as @_
(type variables) instead of Unknown, allowing ParamSpec inference to work
correctly.

jackulau committed:
cargo fmt ran

jackulau committed:
Revert "cargo fmt ran"

This reverts commit dedeeb2c3bdbf40b920630208e6ead002d7706df.

jackulau committed:
Apply cargo fmt to format code

Run cargo fmt to ensure consistent code formatting across the project.

jackulau committed:
Fix generic ParamSpec functions to preserve generic types (#43)

When a generic function or constructor is passed through a ParamSpec-using
function, preserve its generic nature so each call creates fresh type
variables rather than locking types on first call.

- Add ParamSpecValue and Concatenate variants to Forallable enum
- Detect generic params in create_paramspec_value() and wrap in Forall
- Lift Forall wrapper to outer Callable in simplify_mut()
- Handle Forall<ParamSpecValue> in callable.rs to instantiate fresh vars
- Use transform_mut instead of visit_mut to recurse into ClassType's TArgs

yangdanny97 commented on pyrefly/lib/solver/subset.rs:
```diff
@@ -567,6 +571,66 @@ impl<'a, Ans: LookupAnswer> Subset<'a, Ans> {
+                .into_iter()
+                .map(|q| TParam {
+                    quantified: q,
+                    variance: PreInferenceVariance::PInvariant,
```
Is setting all the type vars to invariant safe to do? What happens if the type var is explicitly set to be covariant?

jackulau committed:
Address review feedback on generic ParamSpec PR (#43)

- Fix variance: use PUndefined instead of PInvariant since Quantified
  doesn't store variance info. PUndefined (Bivariant) is safer as it
  won't cause false errors for explicitly covariant/contravariant type vars.
  Added TODO noting proper fix would track variance through fresh_quantified.

- Reduce code duplication: use existing param.as_type() helper instead
  of manual match statements in both solver.rs and subset.rs.

- Remove unused Param import from solver.rs.

NathanTempest commented:
Closing this PR as this has not seen activity in quite a while. Will re-open if the user responds with a follow-up.
